### ISSUE TITLE
False-positives due to renaming re-exports + name collisions in the same crate

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#202]
> Our `cargo semver-checks` CI steps currently fail for `libp2p-dcutr` and `libp2p-relay`, even though neither of them merged any changes since the last [`libp2p` `v0.50.0` release](https://github.com/libp2p/rust-libp2p/releases/tag/v0.50.0).
> 
> - `libp2p-dcutr` https://github.com/libp2p/rust-libp2p/actions/runs/3576177811/jobs/6013685638
> - `libp2p-relay` https://github.com/libp2p/rust-libp2p/actions/runs/3576177811/jobs/6013686091

_Originally posted by @mxinden in https://github.com/libp2p/rust-libp2p/issues/2647#issuecomment-1335189994_

I was able to replicate the issue locally. The generated rustdoc for `libp2p-dcutr` at first glance doesn't seem to have anything out of place. I spot-checked one of the enums and variants that reported an error, and they were present in both the baseline and current rustdoc with identical IDs and reasonable-looking data.

This issue needs further triage.

**Comment by obi1kenobi** [comment]
Unfortunately, I found that this is because of a combination of renaming re-exports causing confusion between multiple enums with the same name. For example, a `pub enum UpgradeError` exists in all of:
```
protocols/dcutr/src/protocol/outbound.rs
protocols/dcutr/src/protocol/inbound.rs
protocols/dcutr/src/behavior.rs
```
The false-positives happen because the lint tries to match up all the variants across different enums, which obviously isn't going to do the right thing.

`cargo-semver-checks` doesn't currently handle re-export renames (and re-exports are already tricky because of the possibility of having infinitely many re-exported names for each item: https://github.com/obi1kenobi/cargo-semver-checks/issues/152 ), so this isn't a trivial fix.

That said, it's still important to fix and this should be sufficient motivation to come up with a permanent solution to the re-exports problem :)

**Comment by obi1kenobi** [comment]
It's tempting to use the `canonical_path` edge instead of `importable_path` edge in the affected lints. Right now, this wouldn't work, since re-exported items do not currently have a `canonical_path` available, since that path uses the rustdoc index information which currently does not include re-exported items.

Relevant rustdoc issues:
https://github.com/rust-lang/rust/issues/93522
https://github.com/rust-lang/rust/issues/99513
